File: Java Project/Iteration_2/src/Transcript.java
package Iteration_2.src;


import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;

public class Transcript {
    // courseGrades is haspMap of course object and grades in integer
    private HashMap<String, String> courseGrades;
    // Student object
    //private Student student;
    // Gpa is a float
    private float gpa;
    // takenCredit is an integer
    private int takenCredit;
    // takenCourses is an arraylist of course object
    private ArrayList<CompulsoryCourse> takenCourses;
    // not taken courses is an arraylist of course object
    private ArrayList<CompulsoryCourse> notTakenCourses;

    // constructor for Transcript takes courseGrade,Student object, gpa,
    // takenCredit, takenCourses, notTakenCourses


    /*public Transcript(HashMap<String, Float> courseGrades, Student student, float gpa, int takenCredit, ArrayList<CompulsoryCourse> takenCourses, ArrayList<CompulsoryCourse> notTakenCourses) {
        this.courseGrades = courseGrades;
        //this.student = student;
        this.gpa = gpa;
        this.takenCredit = takenCredit;
        this.takenCourses = takenCourses;
        this.notTakenCourses = notTakenCourses;
    }*/
    public Transcript(HashMap<String, String> courseGrades, float gpa, int takenCredit, ArrayList<CompulsoryCourse> takenCourses, ArrayList<CompulsoryCourse> notTakenCourses) {
        this.courseGrades = courseGrades;
        this.gpa = gpa;
        this.takenCredit = takenCredit;
        this.takenCourses = takenCourses;
        this.notTakenCourses = notTakenCourses;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    // constructor for Transcript takes nothing as a parameter
    public Transcript() {
        this.courseGrades = new HashMap<>();
        //this.student = new Student();
        this.gpa = 0;
        this.takenCredit = 0;
        this.takenCourses = new ArrayList<>();
        this.notTakenCourses = new ArrayList<>();
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    // set method for Transcript takes courseGrade,Student object, gpa, takenCredit,
    // takenCourses, notTakenCourses
    public void setTranscript(HashMap<String, String> courseGrades, float gpa, int takenCredit,
                              ArrayList<CompulsoryCourse> takenCourses, ArrayList<CompulsoryCourse> notTakenCourses) {
        this.courseGrades = courseGrades;

        this.gpa = gpa;
        this.takenCredit = takenCredit;
        this.takenCourses = takenCourses;
        this.notTakenCourses = notTakenCourses;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();

    }

    public void addTakenCourse(ArrayList<CompulsoryCourse> courses) {
        this.takenCourses.addAll(courses);
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }
    public void addTakenCourseOne(CompulsoryCourse course) {
        this.takenCourses.add(course);
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }
    public void removeCourse(ArrayList<CompulsoryCourse> courses) {
        this.notTakenCourses.remove(courses);
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }
    public  void  addCourseGrade(HashMap<String,String> courseGrades){
        this.courseGrades.putAll(courseGrades);
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }


    public HashMap<String, String> getCourseGrades() {
        return courseGrades;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    public void setCourseGrades(HashMap<String, String> courseGrades) {
        this.courseGrades = courseGrades;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    /*public Student getStudent() {
        return student;
    }

    public void setStudent(Student student) {
        this.student = student;
    }*/

    public float getGpa() {
        return gpa;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    public void setGpa(float gpa) {
        this.gpa = gpa;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    public int getTakenCredit() {
        return takenCredit;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    public void setTakenCredit(int takenCredit) {
        this.takenCredit = takenCredit;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    public ArrayList<CompulsoryCourse> getTakenCourses() {
        return takenCourses;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    public void setTakenCourses(ArrayList<CompulsoryCourse> takenCourses) {
        this.takenCourses = takenCourses;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    public ArrayList<CompulsoryCourse> getNotTakenCourses() {
        return notTakenCourses;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    public void setNotTakenCourses(ArrayList<CompulsoryCourse> notTakenCourses) {
        this.notTakenCourses = notTakenCourses;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    public float calculateGpa() {
        float gpa = 0f;
        int totalCredit = 0;
        for (int j = 0; j < this.getTakenCourses().size(); j++) {
            String letterGrade = this.getCourseGrades().get(this.getTakenCourses().get(j).getCourseCode());
            if (letterGrade == "AA")
                gpa += this.getTakenCourses().get(j).getCredit() * 4;
            else if (letterGrade == "BA")
                gpa += this.getTakenCourses().get(j).getCredit() * 3.5;
            else if (letterGrade == "BB")
                gpa += this.getTakenCourses().get(j).getCredit() * 3;
            else if (letterGrade == "CB")
                gpa += this.getTakenCourses().get(j).getCredit() * 2.5;
            else if (letterGrade == "CC")
                gpa += this.getTakenCourses().get(j).getCredit() * 2;
            else if (letterGrade == "DC")
                gpa += this.getTakenCourses().get(j).getCredit() * 1.5;
            totalCredit += this.getTakenCourses().get(j).getCredit();
        }

        return Math.round((gpa / totalCredit) * 100.0) / 100.0f;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    public HashMap<String, String> randomCourseGrade(ArrayList<CompulsoryCourse> courses) {
        String[] letterGrades = {"AA", "BA", "BB", "CB", "CC", "DC"};
        HashMap<String, String> hashmap = new HashMap<String, String>();

        for (int j = 0; j < courses.size(); j++)
            hashmap.put(courses.get(j).getCourseCode(), letterGrades[new Random().nextInt(letterGrades.length - 1)]);


        return hashmap;
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }

    @Override
    public String toString() {
        return "Transcript{" +
                "courseGrades=" + courseGrades +
                //", student=" + student +
                ", gpa=" + gpa +
                ", takenCredit=" + takenCredit +
                ", takenCourses=" + takenCourses +
                ", notTakenCourses=" + notTakenCourses +
                '}';
        
        Iteration_2.src.LoggingAgent LoggingAgent = new Iteration_2.src.LoggingAgent();
        LoggingAgent.LoggerMethod();
    }
}

File: Python/src/Transcript.py
import random

class Transcript:
    def __init__(self, course_grades, gpa, taken_credit, taken_courses, not_taken_courses):
        self.course_grades = course_grades
        self.gpa = gpa
        self.taken_credit = taken_credit
        self.taken_courses = taken_courses
        self.not_taken_courses = not_taken_courses

    def __init__(self):
        self.course_grades = {}
        self.gpa = 0
        self.taken_credit = 0
        self.taken_courses = []
        self.not_taken_courses = []
    
    def set_transcript(self, course_grades, gpa, taken_credit, taken_courses, not_taken_courses):
        self.course_grades = course_grades
        self.gpa = gpa
        self.taken_credit = taken_credit
        self.taken_courses = taken_courses
        self.not_taken_courses = not_taken_courses

    def add_taken_course(self, courses):
        self.taken_courses.extend(courses)

    def add_taken_course_one(self, course):
        self.taken_courses.append(course)

    def remove_course(self, courses):
        self.not_taken_courses.remove(courses)

    def add_course_grade(self, course_grades):
        self.course_grades.update(course_grades)

    def get_course_grades(self):
        return self.course_grades

    def set_course_grades(self, course_grades):
        self.course_grades = course_grades

    def get_gpa(self):
        return self.gpa

    def set_gpa(self, gpa):
        self.gpa = gpa

    def get_taken_credit(self):
        return self.taken_credit

    def set_taken_credit(self, taken_credit):
        self.taken_credit = taken_credit

    def get_taken_courses(self):
        return self.taken_courses

    def set_taken_courses(self, taken_courses):
        self.taken_courses = taken_courses

    def get_not_taken_courses(self):
        return self.not_taken_courses

    def set_not_taken_courses(self, not_taken_courses):
        self.not_taken_courses = not_taken_courses

    def add_credit(self, added_credit):
        taken_credit = self.get_taken_credit()
        taken_credit += added_credit
        self.set_taken_credit(taken_credit)

    def calculate_gpa(self):
        gpa = 0
        total_credit = 0
        grade_values = {
            "AA": 4,
            "BA": 3.5,
            "BB": 3,
            "CB": 2.5,
            "CC": 2,
            "DC": 1.5,
            "FF": 0
        }
        for j in range(len(self.get_taken_courses())):
            letter_grade = self.get_course_grades().get(self.get_taken_courses()[j].get_course_code())
            gpa += self.get_taken_courses()[j].get_credit() * grade_values.get(letter_grade, 0)
            total_credit += self.get_taken_courses()[j].get_credit()
        
        return round((gpa / total_credit) * 100) / 100


    def random_course_grade(self, courses, prob):
        letter_grades = ["AA", "BA", "BB", "CB", "CC", "DC"]
        hashmap = {}
        for course in courses:
            if random.random() < prob:
                hashmap[course.get_course_code()] = "FF"
            else:
                hashmap[course.get_course_code()] = letter_grades[random.randint(0, len(letter_grades)-1)]
        
        return hashmap

    def __str__(self):
        return f"Transcript{{courseGrades={self.course_grades}, gpa={self.gpa}, takenCredit={self.taken_credit}, takenCourses={self.taken_courses}, notTakenCourses={self.not_taken_courses}}}"






